File: packages/types/src/market/job.ts
/**
 * @skillancer/types - Market: Job Types
 * Job posting and management schemas
 */

import { z } from 'zod';
import {
  uuidSchema,
  dateSchema,
  currencyCodeSchema,
  timestampsSchema,
} from '../common/base';

// =============================================================================
// Job Enums
// =============================================================================

/**
 * Job status lifecycle
 */
export const jobStatusSchema = z.enum([
  'DRAFT',
  'OPEN',
  'IN_PROGRESS',
  'COMPLETED',
  'CANCELLED',
  'ON_HOLD',
]);
export type JobStatus = z.infer<typeof jobStatusSchema>;

/**
 * Job payment type
 */
export const jobTypeSchema = z.enum([
  'FIXED_PRICE',
  'HOURLY',
  'RETAINER',
  'MILESTONE',
]);
export type JobType = z.infer<typeof jobTypeSchema>;

/**
 * Job visibility
 */
export const jobVisibilitySchema = z.enum([
  'PUBLIC',
  'PRIVATE',
  'INVITE_ONLY',
  'TENANT_ONLY',
]);
export type JobVisibility = z.infer<typeof jobVisibilitySchema>;

/**
 * Experience level required
 */
export const experienceLevelSchema = z.enum([
  'ENTRY',
  'INTERMEDIATE',
  'EXPERT',
]);
export type ExperienceLevel = z.infer<typeof experienceLevelSchema>;

/**
 * Project duration estimate
 */
export const projectDurationSchema = z.enum([
  'LESS_THAN_WEEK',
  'ONE_TO_FOUR_WEEKS',
  'ONE_TO_THREE_MONTHS',
  'THREE_TO_SIX_MONTHS',
  'MORE_THAN_SIX_MONTHS',
  'ONGOING',
]);
export type ProjectDuration = z.infer<typeof projectDurationSchema>;

// =============================================================================
// Job Schema
// =============================================================================

/**
 * Job skill requirement
 */
export const jobSkillSchema = z.object({
  name: z.string().min(1).max(100),
  required: z.boolean().default(true),
  yearsRequired: z.number().int().nonnegative().optional(),
});
export type JobSkill = z.infer<typeof jobSkillSchema>;

/**
 * Job attachment
 */
export const jobAttachmentSchema = z.object({
  id: uuidSchema,
  name: z.string().min(1).max(255),
  url: z.string().url(),
  mimeType: z.string(),
  size: z.number().int().positive(),
});
export type JobAttachment = z.infer<typeof jobAttachmentSchema>;

/**
 * Complete job schema
 */
export const jobSchema = z.object({
  id: uuidSchema,
  tenantId: uuidSchema.optional(),
  clientUserId: uuidSchema,
  
  // Basic info
  title: z.string().min(10).max(200),
  description: z.string().min(50).max(10000),
  shortDescription: z.string().max(500).optional(),
  
  // Skills & requirements
  skills: z.array(jobSkillSchema).min(1).max(15),
  experienceLevel: experienceLevelSchema,
  
  // Pricing
  type: jobTypeSchema,
  budgetMin: z.number().positive(),
  budgetMax: z.number().positive(),
  currency: currencyCodeSchema.default('USD'),
  
  // Timeline
  duration: projectDurationSchema.optional(),
  timelineDays: z.number().int().positive().optional(),
  startDate: dateSchema.optional(),
  deadline: dateSchema.optional(),
  
  // Status & visibility
  status: jobStatusSchema,
  visibility: jobVisibilitySchema.default('PUBLIC'),
  
  // SkillPod integration
  secureMode: z.boolean().default(false),
  complianceRequirements: z.array(z.string()).optional(),
  requiredCertifications: z.array(z.string()).optional(),
  
  // Files
  attachments: z.array(jobAttachmentSchema).optional(),
  
  // Metadata
  category: z.string().optional(),
  subcategory: z.string().optional(),
  tags: z.array(z.string()).max(10).optional(),
  
  // Stats
  viewCount: z.number().int().nonnegative().default(0),
  bidCount: z.number().int().nonnegative().default(0),
  
  // Timestamps
  publishedAt: dateSchema.optional(),
  closesAt: dateSchema.optional(),
  ...timestampsSchema.shape,
}).refine((data) => data.budgetMin <= data.budgetMax, {
  message: 'Minimum budget must be less than or equal to maximum budget',
  path: ['budgetMax'],
});
export type Job = z.infer<typeof jobSchema>;

// =============================================================================
// Job CRUD Schemas
// =============================================================================

/**
 * Base job data for create/update
 */
const baseJobDataSchema = z.object({
  title: z.string().min(10).max(200),
  description: z.string().min(50).max(10000),
  shortDescription: z.string().max(500).optional(),
  skills: z.array(jobSkillSchema).min(1).max(15),
  experienceLevel: experienceLevelSchema,
  type: jobTypeSchema,
  budgetMin: z.number().positive(),
  budgetMax: z.number().positive(),
  currency: currencyCodeSchema.default('USD'),
  duration: projectDurationSchema.optional(),
  timelineDays: z.number().int().positive().optional(),
  startDate: dateSchema.optional(),
  deadline: dateSchema.optional(),
  visibility: jobVisibilitySchema.default('PUBLIC'),
  secureMode: z.boolean().default(false),
  complianceRequirements: z.array(z.string()).optional(),
  category: z.string().optional(),
  subcategory: z.string().optional(),
  tags: z.array(z.string()).max(10).optional(),
});

/**
 * Create job input
 */
export const createJobSchema = baseJobDataSchema.extend({
  status: z.enum(['DRAFT', 'OPEN']).default('DRAFT'),
}).refine((data) => data.budgetMin <= data.budgetMax, {
  message: 'Minimum budget must be less than or equal to maximum budget',
  path: ['budgetMax'],
});
export type CreateJob = z.infer<typeof createJobSchema>;

/**
 * Update job input
 */
export const updateJobSchema = baseJobDataSchema.partial();
export type UpdateJob = z.infer<typeof updateJobSchema>;

/**
 * Job search/filter parameters
 */
export const jobFilterSchema = z.object({
  query: z.string().optional(),
  skills: z.array(z.string()).optional(),
  type: z.array(jobTypeSchema).optional(),
  experienceLevel: z.array(experienceLevelSchema).optional(),
  budgetMin: z.number().positive().optional(),
  budgetMax: z.number().positive().optional(),
  currency: currencyCodeSchema.optional(),
  duration: z.array(projectDurationSchema).optional(),
  secureMode: z.boolean().optional(),
  category: z.string().optional(),
  clientUserId: uuidSchema.optional(),
  status: z.array(jobStatusSchema).optional(),
});
export type JobFilter = z.infer<typeof jobFilterSchema>;

File: apps/mobile/lib/features/jobs/domain/models/job.dart
import 'package:equatable/equatable.dart';

/// Budget type enum
enum BudgetType {
  fixed,
  hourly;

  String get displayName => switch (this) {
        BudgetType.fixed => 'Fixed Price',
        BudgetType.hourly => 'Hourly',
      };
}

/// Experience level enum
enum ExperienceLevel {
  entry,
  intermediate,
  expert;

  String get displayName => switch (this) {
        ExperienceLevel.entry => 'Entry Level',
        ExperienceLevel.intermediate => 'Intermediate',
        ExperienceLevel.expert => 'Expert',
      };
}

/// Project duration enum
enum ProjectDuration {
  lessThanOneMonth,
  oneToThreeMonths,
  threeToSixMonths,
  moreThanSixMonths;

  String get displayName => switch (this) {
        ProjectDuration.lessThanOneMonth => 'Less than 1 month',
        ProjectDuration.oneToThreeMonths => '1-3 months',
        ProjectDuration.threeToSixMonths => '3-6 months',
        ProjectDuration.moreThanSixMonths => 'More than 6 months',
      };
}

/// Job model
class Job extends Equatable {
  final String id;
  final String title;
  final String description;
  final String clientName;
  final String? clientAvatarUrl;
  final double budget;
  final BudgetType budgetType;
  final List<String> skills;
  final DateTime postedAt;
  final int proposalCount;
  final ExperienceLevel experienceLevel;
  final ProjectDuration projectDuration;
  final bool isRemote;
  final String? location;
  final int? smartMatchScore;
  final String? category;

  const Job({
    required this.id,
    required this.title,
    required this.description,
    required this.clientName,
    this.clientAvatarUrl,
    required this.budget,
    required this.budgetType,
    required this.skills,
    required this.postedAt,
    required this.proposalCount,
    required this.experienceLevel,
    required this.projectDuration,
    required this.isRemote,
    this.location,
    this.smartMatchScore,
    this.category,
  });

  String get budgetDisplay {
    if (budgetType == BudgetType.hourly) {
      return '\$${budget.toStringAsFixed(0)}/hr';
    }
    return '\$${budget.toStringAsFixed(0)}';
  }

  factory Job.fromJson(Map<String, dynamic> json) {
    return Job(
      id: json['id'] as String,
      title: json['title'] as String,
      description: json['description'] as String,
      clientName: json['clientName'] as String,
      clientAvatarUrl: json['clientAvatarUrl'] as String?,
      budget: (json['budget'] as num).toDouble(),
      budgetType: BudgetType.values.firstWhere(
        (b) => b.name == json['budgetType'],
        orElse: () => BudgetType.fixed,
      ),
      skills: List<String>.from(json['skills'] as List),
      postedAt: DateTime.parse(json['postedAt'] as String),
      proposalCount: json['proposalCount'] as int? ?? 0,
      experienceLevel: ExperienceLevel.values.firstWhere(
        (e) => e.name == json['experienceLevel'],
        orElse: () => ExperienceLevel.intermediate,
      ),
      projectDuration: ProjectDuration.values.firstWhere(
        (d) => d.name == json['projectDuration'],
        orElse: () => ProjectDuration.oneToThreeMonths,
      ),
      isRemote: json['isRemote'] as bool? ?? true,
      location: json['location'] as String?,
      smartMatchScore: json['smartMatchScore'] as int?,
      category: json['category'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'clientName': clientName,
      'clientAvatarUrl': clientAvatarUrl,
      'budget': budget,
      'budgetType': budgetType.name,
      'skills': skills,
      'postedAt': postedAt.toIso8601String(),
      'proposalCount': proposalCount,
      'experienceLevel': experienceLevel.name,
      'projectDuration': projectDuration.name,
      'isRemote': isRemote,
      'location': location,
      'smartMatchScore': smartMatchScore,
      'category': category,
    };
  }

  @override
  List<Object?> get props => [id, title, budget, postedAt];
}
